hermes/crates/cardano-chain-follower/src/lib.rs

```diff
@@ -1,6 +1,213 @@
+
+        Ok(block)
+    }
+}
+
+/// Enum of possible Cardano networks.
+pub enum Network {
+    /// Cardano mainnet network.
+    Mainnet,
+    /// Cardano pre-production network.
+    Preprod,
+    /// Cardano preview network.
+    Preview,
+    /// Cardano testnet network.
+    Testnet,
+}
+
+impl From<Network> for u64 {
+    fn from(network: Network) -> Self {
+        match network {
+            Network::Mainnet => MAINNET_MAGIC,
+            Network::Preprod => PRE_PRODUCTION_MAGIC,
+            Network::Preview => PREVIEW_MAGIC,
+            Network::Testnet => TESTNET_MAGIC,
+        }
+    }
+}
+
+/// Enum of possible types of clients used by the follower.
+enum Client {
+    /// Pallas node-to-node client.
+    N2n(PeerClient),
+    /// Pallas node-to-client client.
+    N2c(NodeClient),
+}
+
+/// Enum of chain updates received by the follower.
+pub enum ChainUpdate {
+    /// New block inserted on chain.
+    Block(MultiEraBlockData),
+    /// Chain rollback to the given point.
+    Rollback(Point),
+}
+
+/// Cardano chain follower.
+pub struct Follower {
+    /// Client used to get chain data.
+    client: Client,
+}
+
+impl Follower {
+    /// Connects the follower to a producer using the node-to-node protocol.
+    ///
+    /// # Arguments
+    ///
+    /// * `address`: Address of the node to connect to.
+    /// * `network`: The [Network] the client is assuming it's connecting to.
+    ///
+    /// # Errors
+    ///
+    /// Returns Err if the connection could not be estabilished.
+    pub async fn connect_n2n(address: &str, network: Network) -> Result<Self> {
+        let client = Client::N2n(
+            PeerClient::connect(address, network.into())
+                .await
+                .map_err(Box::new)?,
+        );
+
+        Ok(Self { client })
+    }
+
+    /// Connects the follower to a producer using the node-to-client protocol.
+    ///
+    /// # Arguments
+    ///
+    /// * `path`: Path to the UDS to use for node communication.
+    /// * `network`: The [Network] the client is assuming it's connecting to.
+    ///
+    /// # Errors
+    ///
+    /// Returns Err if the connection could not be estabilished.
+    #[cfg(unix)]
+    pub async fn connect_n2c<P>(path: P, network: Network) -> Result<Self>
+    where
+        P: AsRef<std::path::Path>,
+    {
+        let client = Client::N2c(
+            NodeClient::connect(path, network.into())
+                .await
+                .map_err(Box::new)?,
+        );
+
+        Ok(Follower { client })
+    }
+
+    /// Set the follower's chain read-pointer. Returns None if the point was
+    /// not found on the chain.
+    ///
+    /// # Arguments
+    ///
+    /// * `at`: Point at which to set the read-pointer.
+    ///
+    /// # Errors
+    ///
+    /// Returns Err if something went wrong while communicating with the producer.
+    pub async fn set_read_pointer<P>(&mut self, _at: P) -> Result<Option<Point>>
```

Comment:
I think that would make the crate's API feel weird for someone who's using it because of the mixed functionalities. e.g. What if I want to just set the read pointer?